src/Capability/State/Internal/Class.hs

```diff
@@ -102,24 +108,44 @@ gets f = do
   pure (f s)
 {-# INLINE gets #-}
 
--- | Execute the given state action on a sub-component of the current state
--- as defined by the given transformer @t@.
+-- | Execute the given state action on a sub-component of the current state as
+-- defined by the given transformer @t@. The set of retained capabilities must
+-- be passed as an extra type parameter. If no capabilities are required,
+-- 'Capabilities.Constraints.None' can be used.
 --
--- Example:
+-- Examples:
 --
--- > zoom @"foobar" @"foo" @(Field "foo" "foobar") foo
+-- > zoom @"foobar" @"foo" @(Field "foo" "foobar") @None foo
 -- >   :: HasState "foobar" FooBar m => m ()
 -- >
 -- > foo :: HasState "foo" Int m => m ()
 -- > data FooBar = FooBar { foo :: Int, bar :: String }
 --
+-- > zoom @"foobar" @"foo" @(Field "foo" "foobar") @('[MonadIO]) bar
+-- >   :: (MonadIO m, HasState "foobar" FooBar m) => m ()
+-- >
+-- > foo :: HasState "foo" Int m => m ()
+-- > bar :: (MonadIO m, HasState "foo" Int m) => m ()
+-- > data FooBar = FooBar { foo :: Int, bar :: String }
+--
 -- This function is experimental and subject to change.
 -- See <https://github.com/tweag/capability/issues/46>.
-zoom :: forall outertag innertag t outer inner m a.
-  ( forall x. Coercible (t m x) (m x)
+zoom :: forall outertag innertag t (cs :: [(* -> *) -> Constraint]) outer inner m a.
+  ( forall x. Coercible (m x) (t m x)
   , forall m'. HasState outertag outer m'
     => HasState innertag inner (t m')
-  , HasState outertag outer m )
-  => (forall m'. HasState innertag inner m' => m' a) -> m a
-zoom m = coerce @(t m a) m
+  , HasState outertag outer m
+  , All cs m )
+  => (forall m'. All (HasState innertag inner ': cs) m' => m' a) -> m a
+zoom action =
+  let constraintsDict =
+        -- Note: this use of 'unsafeCoerce' should be safe thanks the Coercible
```

Comment:
```suggestion
        -- Note: this use of 'unsafeCoerce' is safe thanks the Coercible
```

It is also important that you are passing the dictionary to a function of type `forall m' … m' a`, otherwise you could potentially have conflicting dictionaries. This is (in part) why `Dict`'s has a nominal role. (there is also a such existentially quantified type in the reflection package for the same reason).